mllib/src/main/scala/org/apache/spark/ml/tree/Node.scala

```diff
@@ -287,6 +291,34 @@ private[tree] class LearningNode(
     }
   }
 
+  /**
+   * @return true iff the node is a leaf.
+   */
+  private def isLeafNode(): Boolean = leftChild.isEmpty && rightChild.isEmpty
+
+  // the set of (leaf) predictions appearing in the subtree rooted at the given node.
+  private lazy val leafPredictions: Set[Double] = {
```

Comment:
This will store a potentially very large collection at each node. For deep regression trees the storage cost could be quite large. We can accomplish the same thing without storing them:

```scala
  def toNode: Node = {

    // convert to an inner node only when:
    //  -) the node is not a leaf, and
    //  -) the subtree rooted at this node cannot be replaced by a single leaf
    //     (i.e., there at least two different leaf predictions appear in the subtree)
    if (!isLeafNode) {
      assert(leftChild.nonEmpty && rightChild.nonEmpty && split.nonEmpty && stats != null,
        "Unknown error during Decision Tree learning.  Could not convert LearningNode to Node.")
      (leftChild.get.toNode, rightChild.get.toNode) match {
        case (l: LeafNode, r: LeafNode) if l.prediction == r.prediction =>
          new LeafNode(l.prediction, stats.impurity, stats.impurityCalculator)
        case (l, r) =>
          new InternalNode(stats.impurityCalculator.predict, stats.impurity, stats.gain,
            l, r, split.get, stats.impurityCalculator)
      }
    } else {
      if (stats.valid) {
        new LeafNode(stats.impurityCalculator.predict, stats.impurity,
          stats.impurityCalculator)
      } else {
        // Here we want to keep same behavior with the old mllib.DecisionTreeModel
        new LeafNode(stats.impurityCalculator.predict, -1.0, stats.impurityCalculator)
      }
    }
  }
```

Reply:
It's only stored during training though. I agree it could turn into a problem, but, wondered how many distinct predictions there could be? in the case of regression, maybe a lot, hm.

Doesn't this only prune cases where a node has two leaf-node children with the same prediction? 

We should be able to prune much more than that. You could do so in multiple passes. The other way is roughly what's done here, to remember the predictions in each learning node.

Hm, yeah now I'm wondering about the regression case here.

Reply:
What other cases are there that I'm missing? For one thing, the current tests pass with this change. 

This will prune a subtree of arbitrary depth that has a single distinct prediction in its leaf nodes.

Reply:
Imagine a full binary tree of 2 levels (7 nodes) with all identical predictions. It can be pruned down to one node, the root. But I think this logic would only prune the lowest level of nodes. Right? you're matching on nodes that have `LeafNode` children only.

Reply:
This is a recursive method. It essentially does this:

```scala
def prune(node: Node): Node = {
  val left = prune(node.left)
  val right = prune(node.right)
  if (left.isLeaf && right.isLeaf && left.prediction == right.prediction) {
    new LeafNode
  } else {
    new InternalNode
  }
}
```

It starts pruning at the bottom and then works its way up.

Reply:
Oh right, you call `toNode` in here. My doubt there was that `toNode` gets called on each node to turn learning nodes into final ones. Wouldn't this recompute the subtrees over and over? I might have misread how this part works though.

Reply:
The variant proposed by @sethah looks complete to me, and it was indeed not necessary to store the predictions.

If I got right the counterexample of @srowen, the new code would behave on it in this way:

(input tree)
1: InnerNode (not built yet)
--1.1: InnerNode (not built yet)
----1.1.1 (label x): LeafNode (not built yet)
----1.1.2 (label x): LeafNode (not built yet)
--1.2: InnerNode (not built yet)
----1.2.1 (label x): LeafNode (not built yet)
----1.2.2 (label x): LeafNode (not built yet)

toNode on 1.1.1 and 1.1.2 builds two LeafNode, so when recursion come back to the context of 1.1, we have the following situation:

1: InnerNode (not built yet)
--1.1: InnerNode (not built yet)
----1.1.1 (label x): LeafNode (not built yet)
----1.1.2 (label x): LeafNode (not built yet)
--1.2: InnerNode (not built yet)
----1.2.1 (label x): LeafNode (not built yet)
----1.2.2 (label x): LeafNode (not built yet)

Therefore, the first case is triggered, and 1.1 becomes a LeafNode, giving us this situation:

1: InnerNode (not built yet)
--1.1 (label x): LeafNode
--1.2: InnerNode (not built yet)
----1.2.1 (label x): LeafNode (not built yet)
----1.2.2 (label x): LeafNode (not built yet)

Then, toNode is called on 1.2, and similarly to what happened to subtree rooted at 1.1 we have a match on the first case, resulting in this situation:

1: InnerNode (not built yet)
--1.1 (label x): LeafNode 
--1.2 (label x): LeafNode 

Now toNode on 1.1 and 1.2 has built two LeafNode, and the first case matches also in the context of node 1, that is finally translated into a single LeafNode, as expected.

Reply:
The recursion pattern is the same as before, since it still just calls `toNode` on left and right when they are defined.

Reply:
Got it, I see how it works now. `toNode` is called recursively once from the root, not serially on each node, and that's how it always worked. Right, this is much simpler then.